pjsip/src/pjsip-ua/sip_inv.c

```diff
@@ -3066,6 +3068,107 @@ PJ_DEF(pj_status_t) pjsip_inv_process_redirect( pjsip_inv_session *inv,
+                param = uri->header_param.next;
+                while (param != &uri->header_param) {
+                    const pjsip_hdr_e hdr_types[] =
+                    {
+                    /* We should reject the following headers as
+                     * they can potentially be malicious or misleading.
+                     */
+                    PJSIP_H_ACCEPT, PJSIP_H_ACCEPT_ENCODING_UNIMP,
+                    PJSIP_H_ACCEPT_LANGUAGE_UNIMP, PJSIP_H_ALLOW,
+                    PJSIP_H_CALL_ID, PJSIP_H_CONTACT, PJSIP_H_CSEQ,
+                    PJSIP_H_FROM, PJSIP_H_ORGANIZATION_UNIMP,
+                    PJSIP_H_RECORD_ROUTE, PJSIP_H_ROUTE,
+                    PJSIP_H_SUPPORTED, PJSIP_H_TO,
+                    PJSIP_H_USER_AGENT_UNIMP, PJSIP_H_VIA,
+                    /* For future use, we should verify the content
+                     * of the following headers, but these headers
+                     * are currently unimplemented in PJSIP.
+                     */
+                    // PJSIP_H_CONTENT_DISPOSITION_UNIMP,
+                    // PJSIP_H_CONTENT_ENCODING_UNIMP,
+                    // PJSIP_H_CONTENT_LANGUAGE_UNIMP,
+                    // PJSIP_H_DATE_UNIMP, PJSIP_H_MIME_VERSION_UNIMP,
+                    // PJSIP_H_TIMESTAMP_UNIMP,
+                    /* We opt to ignore the following headers as
+                     * they will be automatically populated by PJSIP.
+                     */
+                    PJSIP_H_CONTENT_LENGTH, PJSIP_H_CONTENT_TYPE
+                    };
+                    pjsip_generic_string_hdr *hdr;
+                    unsigned i;
+                    pj_bool_t found = PJ_FALSE;
+
+                    /* Check if we should reject/ignore the header. */
+                    for (i = 0; i < sizeof(hdr_types)/sizeof(pjsip_hdr_e); i++)
+                    {
+                        if (!pj_stricmp2(&param->name,
+                                         pjsip_hdr_names[hdr_types[i]].name))
+                        {
+                            found = PJ_TRUE;
+                            break;
+                        }
+                    }
+                    if (found) {
+                        PJ_LOG(4, (THIS_FILE, "Redirection header %.*s "
+                                   "rejected/ignored",
+                                   (int)param->name.slen, param->name.ptr));
+                        param = param->next;
+                        continue;
+                    }
+
+                    /* If the header already exists, we choose to
+                     * overwrite it.
+                     */
+                    hdr = pjsip_msg_find_hdr_by_name(tdata->msg, &param->name,
+                                                     NULL);
+                    if (hdr) {
+                        pjsip_msg_find_remove_hdr(tdata->msg,
+                                                  PJSIP_H_OTHER, hdr);
```

Comment:
I'm not sure what's the correct processing for Authorization and Proxy-Authorization